cmd/contour/serve.go

```diff
@@ -524,32 +514,32 @@ func (s *Server) setupRateLimitService(contourConfiguration contour_api_v1alpha1
 		return nil
 	}
 
-	namespacedName := &types.NamespacedName{
+	// ensure the specified ExtensionService exists
+	extensionSvc := &contour_api_v1alpha1.ExtensionService{}
+	key := client.ObjectKey{
 		Namespace: contourConfiguration.RateLimitService.ExtensionService.Namespace,
 		Name:      contourConfiguration.RateLimitService.ExtensionService.Name,
 	}
-	client := s.clients.DynamicClient().Resource(contour_api_v1alpha1.ExtensionServiceGVR).Namespace(namespacedName.Namespace)
 
-	// ensure the specified ExtensionService exists
-	res, err := client.Get(context.Background(), namespacedName.Name, metav1.GetOptions{})
-	if err != nil {
-		return fmt.Errorf("error getting rate limit extension service %s: %v", namespacedName, err)
-	}
-	var extensionSvc contour_api_v1alpha1.ExtensionService
-	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(res.Object, &extensionSvc); err != nil {
-		return fmt.Errorf("error converting rate limit extension service %s: %v", namespacedName, err)
+	// Using GetAPIReader() here because the manager's caches won't be started yet,
+	// so reads from the manager's client (which uses the caches for reads) will fail.
+	if err := s.mgr.GetAPIReader().Get(context.Background(), key, extensionSvc); err != nil {
```

Comment:
It would be neat to allow Contour to start and pick up on the extension service if its created later on.

Reply:
Filed #4155.